test/e2e/network_policy.go

```diff
@@ -0,0 +1,502 @@
+			ObjectMeta: api.ObjectMeta{
+				Name: "allow-client-a-via-pod-selector",
+			},
+			Spec: extensions.NetworkPolicySpec{
+				// Apply this policy to the Server
+				PodSelector: unversioned.LabelSelector{
+					MatchLabels: map[string]string{
+						"pod-name": serverPod.Name,
+					},
+				},
+				// Allow traffic only from client-a
+				Ingress: []extensions.NetworkPolicyIngressRule{{
+					From: []extensions.NetworkPolicyPeer{{
+						PodSelector: &unversioned.LabelSelector{
+							MatchLabels: map[string]string{
+								"pod-name": clientAName,
+							},
+						},
+					}},
+				}},
+			},
+		})
+		Expect(err).NotTo(HaveOccurred())
+		defer func() {
+			By("Cleaning up the policy.")
+			if err := f.Client.NetworkPolicies(ns.Name).Delete(policy.Name, nil); err != nil {
+				framework.Failf("unable to delete policy %v: %v", policy.Name, err)
+			}
+		}()
+
+		By("Creating client-a which should be able to contact the server.")
+		podClientA := createClientPod(f, ns, clientAName, service, listeningPort1)
+		defer func() {
+			By("Cleaning up the pod client-a")
+			if err := f.Client.Pods(ns.Name).Delete(podClientA.Name, nil); err != nil {
+				framework.Failf("unable to delete pod %v: %v", podClientA.Name, err)
+			}
+		}()
+
+		framework.Logf("Waiting for client-a to come up.")
+		err = framework.WaitForPodRunningInNamespace(f.Client, podClientA)
+		Expect(err).NotTo(HaveOccurred(), "waiting for Client-A to run")
+		framework.Logf("Waiting for client-a to complete.")
+		err = framework.WaitForPodSuccessInNamespace(f.Client, podClientA.Name, ns.Name)
+		Expect(err).NotTo(HaveOccurred(), "checking client-a could communicate with server.")
+
+		// Create a pod with name 'client-b', which will attempt to comunicate with the server,
+		// but should not be able to due to the label-based policy.
+		By("Creating client-b which should *not* be able to contact the server.")
+		podClientB := createClientPod(f, ns, clientBName, service, listeningPort1)
+		defer func() {
+			By("Cleaning up the pod Server A")
+			if err := f.Client.Pods(ns.Name).Delete(podClientB.Name, nil); err != nil {
+				framework.Failf("unable to delete pod %v: %v", podClientB.Name, err)
+			}
+		}()
+		framework.Logf("Waiting for client-b to come up.")
+		err = framework.WaitForPodRunningInNamespace(f.Client, podClientB)
+		Expect(err).NotTo(HaveOccurred(), "waiting for Client-B to run")
+		framework.Logf("Waiting for client-b to complete.")
+		err = framework.WaitForPodSuccessInNamespace(f.Client, podClientB.Name, ns.Name)
+		Expect(err).To(HaveOccurred(), "checking client-b could not communicate with server")
+	})
+
+	It("should enforce policy based on Ports [Feature:NetworkPolicy]", func() {
+		ns := f.Namespace
+		setNamespaceIsolation(f, ns, "DefaultDeny")
+
+		// Create Server with Service
+		By("Creating a simple server.")
+		serverPod, service := createServerPodAndService(f, ns, serverName, []int{listeningPort1, listeningPort2})
```

Comment:
This is the only test case that passes multiple ports to `createServerPodAndService`, but nothing tests that `createServerPodAndService` actually works correctly in that case. (If there is a bug in `createServerPodAndService` such that it only opens the first port, this test would still pass.) And then if that happened, then the test would *always* pass, even if the NetworkPolicy implementation was broken, because it doesn't distinguish client-b failing because of isolation vs client-b failing because of connection refused.

Simple fix: flip the policy around so port2 is allowed and port1 is denied, and test that instead (and assume that `createServerPodAndService` isn't likely to be buggy in a way that would let this pass erroneously)
Better fix: test that both listeningPort1 and listeningPort2 work before applying the NetworkPolicy